x-pack/plugin/inference/src/test/java/org/elasticsearch/xpack/inference/external/elastic/ElasticInferenceServiceResponseHandlerTests.java

```diff
@@ -0,0 +1,119 @@
+import org.elasticsearch.xpack.inference.external.request.Request;
+import org.hamcrest.MatcherAssert;
+
+import java.nio.charset.StandardCharsets;
+
+import static org.hamcrest.Matchers.containsString;
+import static org.hamcrest.core.Is.is;
+import static org.mockito.ArgumentMatchers.anyString;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+
+public class ElasticInferenceServiceResponseHandlerTests extends ESTestCase {
+
+    public void testCheckForFailureStatusCode_DoesNotThrowFor200() {
+        callCheckForFailureStatusCode(200, "id");
+    }
+
+    public void testCheckForFailureStatusCode_ThrowsFor400() {
+        var exception = expectThrows(RetryException.class, () -> callCheckForFailureStatusCode(400, "id"));
+        assertFalse(exception.shouldRetry());
+        MatcherAssert.assertThat(
+            exception.getCause().getMessage(),
+            containsString("Received a bad request status code for request from inference entity id [id] status [400]")
+        );
+        MatcherAssert.assertThat(((ElasticsearchStatusException) exception.getCause()).status(), is(RestStatus.BAD_REQUEST));
+    }
+
+    public void testCheckForFailureStatusCode_ThrowsFor405() {
+        var exception = expectThrows(RetryException.class, () -> callCheckForFailureStatusCode(405, "id"));
+        assertFalse(exception.shouldRetry());
+        MatcherAssert.assertThat(
+            exception.getCause().getMessage(),
+            containsString("Received a method not allowed status code for request from inference entity id [id] status [405]")
+        );
+        MatcherAssert.assertThat(((ElasticsearchStatusException) exception.getCause()).status(), is(RestStatus.METHOD_NOT_ALLOWED));
+    }
+
+    public void testCheckForFailureStatusCode_ThrowsFor413() {
+        var exception = expectThrows(ContentTooLargeException.class, () -> callCheckForFailureStatusCode(413, "id"));
+        assertTrue(exception.shouldRetry());
+        MatcherAssert.assertThat(
+            exception.getCause().getMessage(),
+            containsString("Received a content too large status code for request from inference entity id [id] status [413]")
+        );
+        MatcherAssert.assertThat(((ElasticsearchStatusException) exception.getCause()).status(), is(RestStatus.REQUEST_ENTITY_TOO_LARGE));
+    }
+
+    public void testCheckForFailureStatusCode_ThrowsFor500_WithShouldRetryTrue() {
+        var exception = expectThrows(RetryException.class, () -> callCheckForFailureStatusCode(500, "id"));
+        assertTrue(exception.shouldRetry());
+        MatcherAssert.assertThat(
+            exception.getCause().getMessage(),
+            containsString("Received a server error status code for request from inference entity id [id] status [500]")
+        );
+        MatcherAssert.assertThat(((ElasticsearchStatusException) exception.getCause()).status(), is(RestStatus.BAD_REQUEST));
```

Comment:
Surely this should be `is(RestStatus.INTERNAL_SERVER_ERROR)` for an 500 status code?

The tests pass but I don't understand how

Reply:
Interesting... <500 error codes get converted to 400 [here](https://github.com/elastic/elasticsearch/blob/1163d2e4f957126b7847fd6ee7e6544f860da128/x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/external/http/retry/BaseResponseHandler.java#L89), no idea why. I'm assuming we don't want to expose the specifics of the downstream service.

Maybe @jonathan-buttner can confirm as he is the author of this code?

Reply:
I think because 500 error comes from an external service and we don't want it to appear as an Elasticsearch server error in the logs